### PR TITLE
Fix tcp observes

### PR DESCRIPTION
**Pull request by boaks**
Fix several tcp issues.

Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

### CODE REVIEW

File: californium-core/src/main/java/org/eclipse/californium/core/network/InMemoryMessageExchangeStore.java [hunk 1]
```diff
@@ -362,7 +364,6 @@ public boolean registerOutboundResponse(final Exchange exchange) {
 			throw new IllegalArgumentException("exchange does not contain a response");
 		} else {
 			Response currentResponse = exchange.getCurrentResponse();
-			exchange.retransmitResponse();
```
Comment (boaks) [inline californium-core/src/main/java/org/eclipse/californium/core/network/InMemoryMessageExchangeStore.java]: That's moved to `CoapEndpoint.sendResponse`. 
Potentially changes the UDP processing of notifies on the coap-server (LWM2M device) also. 
All test are passing, so I think this change doesn't break the UDP stuff.

Comment (sbernard31) [reply californium-core/src/main/java/org/eclipse/californium/core/network/InMemoryMessageExchangeStore.java]: I'm not sure to understand how works the exchange.restransmitResponse() function.

But it seems that before we do that for all response and now only for request with observe option, is it intended ?

**Comment by boaks** [comment]
In this cases, it's a "received MID". I f no MID is received, it should not be rejected, because rejection is based on the MID. So it's either malformed or tcp, but both should not be rejected.

**Comment by boaks** [comment]
So, is it OK to merge this PR?

File: californium-core/src/main/java/org/eclipse/californium/core/network/InMemoryMessageExchangeStore.java (hunk 1 above)
Comment (boaks) [reply californium-core/src/main/java/org/eclipse/californium/core/network/InMemoryMessageExchangeStore.java]: For all responses, which are registered as outbound response, so all CON and notifies.
This PR decouples the retransmitResponse for the registerOutboundResponse and use the observe nature of the request to trigger the retransmitResponse.

Comment (sbernard31) [reply californium-core/src/main/java/org/eclipse/californium/core/network/InMemoryMessageExchangeStore.java]: I'm sorry but I didn't get the point at all :'(

Comment (boaks) [reply californium-core/src/main/java/org/eclipse/californium/core/network/InMemoryMessageExchangeStore.java]: > I'm not sure to understand how works the exchange.restransmitResponse() function.

That hasn't changed. The idea was to keep control about the state changes in the exchange. Before I introduced it, a lot of exchange stuff was called several times from several places, which doesn't made it easy to fix the "leaks".

> But it seems that before we do that for all response

For all responses, which are registered as outbound response, so all CON and notifies. See the callers of that `registerOutboundResponse`.
(And for sending responses to duplicate request, but his isn't changed).
 
> and now only for request with observe option, is it intended ?
  
Yes. Usual exchanges only send one response by request. Only observes send more.

Comment (boaks) [reply californium-core/src/main/java/org/eclipse/californium/core/network/InMemoryMessageExchangeStore.java]: So, do you ask, for the function of the `exchange.restransmitResponse()` or for the changes her in this PR according the use of it?

Comment (sbernard31) [reply californium-core/src/main/java/org/eclipse/californium/core/network/InMemoryMessageExchangeStore.java]: That was a generic question about what is the purpose of this function (no specific of this PR)

Comment (boaks) [reply californium-core/src/main/java/org/eclipse/californium/core/network/InMemoryMessageExchangeStore.java]: OK, so does the short hint above help?

Comment (sbernard31) [reply californium-core/src/main/java/org/eclipse/californium/core/network/InMemoryMessageExchangeStore.java]: Nope :'(
I read `exchange.restransmitResponse()` and this do almost nothing.

Comment (boaks) [reply californium-core/src/main/java/org/eclipse/californium/core/network/InMemoryMessageExchangeStore.java]:
```
	/**
	 * Prepare exchange for retransmit a response.
	 * 
	 * @throws IllegalStateException if exchange is not a REMOTE exchange.
	 * @throws ConcurrentModificationException, if not executed within the
	 *             {@link StripedExchangeJob}.
	 */
	public void retransmitResponse() {
		assertOwner();
		if (origin == Origin.REMOTE) {
			caller = null;
			complete.set(false);
		} else {
			throw new IllegalStateException(this + " retransmit on local exchange not allowed!");
		}
	}
```

So usually, for remote exchanges, only one response is assumed. If more responses are intended to be sent, call this. If not called, but more than one reponse is sent, then

```
	private int registerWithMessageId(final Exchange exchange, final Message message) {
		enableStatus = true;
		exchange.assertIncomplete(message);
```
will suppress this unintended re-response. The stricter status monitoring helps to make the house keeping proper.

Comment (boaks) [reply californium-core/src/main/java/org/eclipse/californium/core/network/InMemoryMessageExchangeStore.java]: It seems, we discuss more about the PR, which introduces this exchange-state-monitoring, then about this PR. 
The changes in this PR are quite simple, as explained above. Only deduplication and observes requires to send multiple responses.

Comment (sbernard31) [reply californium-core/src/main/java/org/eclipse/californium/core/network/InMemoryMessageExchangeStore.java]: > It seems, we discuss more about the PR, which introduces this exchange-state-monitoring, then about this PR.

I agree, bu I was not able to understand without that.

And to be honest, I sill don't understand this changes. Anyway go to merge it.

**Pull request merged**